nodecg-io-nanoleaf/extension/nanoleafClient.ts

```diff
@@ -0,0 +1,215 @@
+import { Response } from "node-fetch";
+import { NodeCG } from "nodecg/types/server";
+import { Color, ColoredPanel, PanelEffect } from "./interfaces";
+import { NanoleafQueue } from "./nanoleafQueue";
+import { NanoleafUtils } from "./nanoleafUtils";
+
+export class NanoleafClient implements ServiceClient<NanoleafClient> {
+    // Important: Does only remember colors which were directly set by using setPanelColor(s)
+    private colors: Map<number, Color> = new Map<number, Color>();
+
+    // This queue is used to queue effects
+    private queue: NanoleafQueue = new NanoleafQueue();
+
+    /**
+     * Returns the client-specific effect queue.
+     */
+    getQueue(): NanoleafQueue {
+        return this.queue;
+    }
+
+    getNativeClient(): NanoleafClient {
+        return this; // yolo
+    }
+
+    constructor(private ipAddress: string, private authToken: string, nodecg: NodeCG) {
+        NanoleafUtils.verifyAuthKey(ipAddress, authToken).then((response) => {
+            if (response) {
+                nodecg.log.info("Connected to Nanoleafs successfully.");
+            } else {
+                nodecg.log.error("Unable to connect to Nanoleafs! Please check your credentials!");
+            }
+        });
+    }
+
+    private async callGET(relativePath: string) {
+        return fetch(NanoleafUtils.buildBaseRequestAddress(this.ipAddress, this.authToken) + relativePath, {
+            method: "GET",
+        });
+    }
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    private async callPUT(relativePath: string, body: any) {
+        return fetch(NanoleafUtils.buildBaseRequestAddress(this.ipAddress, this.authToken) + relativePath, {
+            method: "PUT",
+            headers: {
+                "Content-Type": "application/json",
+            },
+            body: JSON.stringify(body),
+        });
+    }
+
+    /**
+     * Returns information about all panels, e.g. available effects, position data, ...
+     */
+    async getAllPanelInfo(): Promise<Response> {
+        return this.callGET("");
+    }
+
+    /**
+     * Returns the IDs of all panels which are connected to the nanoleaf controller
+     * @param sortedByY the IDs are sorted by y level if true, otherwise sorted by x level
+     */
+    async getAllPanelIDs(sortedByY: boolean): Promise<Array<number>> {
+        const response = await this.getAllPanelInfo();
+
+        if (response.status !== 200) {
+            return [];
+        }
+
+        const json = await response.json();
+        // eslint-disable-next-line @typescript-eslint/no-explicit-any
+        const positionData: Array<any> = json.panelLayout?.layout?.positionData;
+        const panels = sortedByY ? positionData.sort((a, b) => a.y - b.y) : positionData.sort((a, b) => a.x - b.x);
+        const panelIDs = panels?.map((entry: { panelId: number }) => entry.panelId);
+        const panelIDsWithoutController = panelIDs.filter((entry: number) => entry !== 0);
```

Comment:
This can be made more type safe by providing a type with all files that were used here. (Obviously not perfect but better than `any`)
```suggestion
        const positionData: Array<{ x: number; y: number; panelId: number }> = json.panelLayout?.layout?.positionData;
        const panels = sortedByY ? positionData.sort((a, b) => a.y - b.y) : positionData.sort((a, b) => a.x - b.x);
        const panelIDs = panels?.map((entry) => entry.panelId);
        const panelIDsWithoutController = panelIDs.filter((entry: number) => entry !== 0);
```